sdk/spring/azure-spring-boot-test-aad/HOW_TO_RUN_AAD_INTEGRATION_TEST_IN_LOCALHOST.md

```diff
@@ -0,0 +1,116 @@
+
+2. After the tenant created successfully. Click `Tenant creation was successful. Click here to navigate to your new tanent: aadIntegrationTest1`.
+![created-a-new-tenant](images/created-a-new-tenant.png)
+
+3. Copy `Tenant ID`. Then we get the value of `AAD_TENANT_ID_1`.
+![copy-tenant-id](images/copy-tenant-id.png)
+
+### `AAD_USER_NAME_1`, `AAD_USER_PASSWORD_1`
+
+1. Create a new user named `user1`. Refs: [Add or delete users using Azure Active Directory][Add or delete users using Azure Active Directory].
+
+2. Copy the user name and initial password. In the following image, user name = `user1@aadIntegrationTest1.onmicrosoft.com`, initial password = `Xabo2448`.
+![create-user1](images/create-user1.png)
+
+3. In Edge, open an InPrivate window to change password for new created user.
+![open-inPrivate-window](images/open-inPrivate-window.png)
+![change-password](images/change-password.png)
+
+4. Now we get the value of `AAD_USER_NAME_1`, `AAD_USER_PASSWORD_1`.
+
+### `AAD_TENANT_ID_2`, `AAD_USER_NAME_2`, `AAD_USER_PASSWORD_2` 
+Get `AAD_TENANT_ID_2`, `AAD_USER_NAME_2`, `AAD_USER_PASSWORD_2` with the same method.
+
+### `AAD_MULTI_TENANT_CLIENT_ID`, `AAD_MULTI_TENANT_CLIENT_SECRET`.
+
+1. Register an application. Refs: [Register an application with the Microsoft identity platform][Register an application with the Microsoft identity platform]
+![create-aad-multitenant-client](images/create-aad-multitenant-client.png)
+
+2. Create secret for `aad-multitenant-client`. Refs: [Create a new application secret][Create a new application secret]
+
+3. Allow implicit flow
+![allow-implicit-flow](images/allow-implicit-flow.png)
+
+4. Add API permissions. Grant admin consent.
+![add-api-permission](images/add-api-permission.png)
+ 
+5. Now we get value of `AAD_MULTI_TENANT_CLIENT_ID`, `AAD_MULTI_TENANT_CLIENT_SECRET`.
+
+### `AAD_SINGLE_TENANT_CLIENT_ID`, `AAD_SINGLE_TENANT_CLIENT_SECRET`
+Get `AAD_SINGLE_TENANT_CLIENT_ID`, `AAD_SINGLE_TENANT_CLIENT_SECRET` with the same method.
+
+### `AAD_SINGLE_TENANT_CLIENT_ID_WITH_ROLE`, `AAD_SINGLE_TENANT_CLIENT_SECRET_WITH_ROLE`
+Get `AAD_SINGLE_TENANT_CLIENT_ID_WITH_ROLE`, `AAD_SINGLE_TENANT_CLIENT_SECRET_WITH_ROLE` with the same method.
+
+More steps needed for this client is to set App roles.
+1. Create roles
+![role-user](images/role-user.png)
+
+2. Assign roles by clicking: "Azure Active Directory" -> "Enterprise Application" -> choose your application -> "Users and groups" -> assign the new roles to user1. (assignment of roles to groups is not available in the free tier of AAD).
+
+### Create groups
+1. Create 2 groups: group1 and group2. Add user1 in group1, and user2 in group2.
+Refs: [Create a basic group and add members using Azure Active Directory][Create a basic group and add members using Azure Active Directory]
+
+2. For user1, we need to create 111 groups: from group1 to group111. It's used in AADGroupsCountIT.
```

Comment:
Can we make this optional? Community contributors will be scared to make contributions.

Reply:
🤣

1. After make this optional, mvn command will fail with test failure.
2. In my understanding, no community contributors will run the test. Only member in our team will run it manually.

Originally, I have a plan to write script to create the 111 groups, but abandoned. Reasons:
1. Write script will cost more time than do it manually.
2. It's not a bad thing to let us be familiar with Azure Portal.

Reply:
OK, I'll add `Optional`